andela/ah-django-unchained
# #162948846 Users should be able to follow each other

**What does this PR do?**
Creates _friends_ application to handle following and unfollowing users

**Description of tasks to be completed**

- Create a Friend model class to handle user relationships
- Create a CustomSerializer to verify the usernames
- Write the views
- Add the routes

**Any background context you want to provide?**

- A user should be able to follow and be followed by other users
- A user can unfollow and be unfollowed by other users
- A user can view his/her followers (or other users followers)
- A user can see who they are following (or who other users are following)

**How should this be manually tested?**

- `git clone https://github.com/andela/ah-django-unchained`
- `cd ah-django-unchained`
- `git checkkout ft-user-follow-unfollow-162948846`
- `python3 -m venv virtual_env_name` to create a virtual environment
- Add a `.env` file as indicated in the ReadMe
- `source .env` to load your environment variables
- `pip install -r requirements.txt`
- Run the migrations `python manage.py makemigrations && python manage.py migrate`

On Postman, register 2 or more users via the endpoint `http://127.0.0.1:8000/api/users/`.

To follow a particular user
- Make a `POST` request on `http://127.0.0.1:8000/api/profiles/<username>/follow` where the username should be of the user you intend to follow, ensure to copy the token of one of the other user and add it as an authorization header in the request. (In all the following screenshots the current signed-in user is `Caren`)
- The response returns the user id of the user being followed.
<img width="970" alt="screenshot 2019-01-18 at 15 20 39" src="https://user-images.githubusercontent.com/32644295/51386794-c6dafb00-1b34-11e9-8f2c-ff59e662b5c3.png">



To view the followers of the user that you just followed,
- Make a `GET` request on `http://127.0.0.1:8000/api/profiles/<username>/followers` where the username should be of the user you just followed.

<img width="968" alt="screenshot 2019-01-18 at 15 22 44" src="https://user-images.githubusercontent.com/32644295/51386865-01dd2e80-1b35-11e9-84c7-807271f15440.png">



To view who you are following,
- Make a `GET` request on `http://127.0.0.1:8000/api/profiles/<username>/following` where the username should be of the current signed in user.
<img width="962" alt="screenshot 2019-01-18 at 15 24 16" src="https://user-images.githubusercontent.com/32644295/51386913-29cc9200-1b35-11e9-9a69-3c672e483019.png">


To unfollow a particular user
- Use url `http://127.0.0.1:8000/api/profiles/<username>/follow` with `DELETE` where the username should be of the user you followed, make sure you copy the token from the registration response and add it as an authorization header in the request
- The response returns the user id of the user being unfollowed.

<img width="968" alt="screenshot 2019-01-18 at 15 16 56" src="https://user-images.githubusercontent.com/32644295/51386654-40beb480-1b34-11e9-8450-4f0a3b9f17fe.png">



**What are the relevant Pivotal Tracker Stories?**
[#162948846](https://www.pivotaltracker.com/story/show/162948846)
